### PR TITLE
feat(x/ecocredit): Put basket server method

### PR DESCRIPTION
**Pull request by technicallyty**
## Description

Implements the `Put` method for baskets.
Tests for `Put` method.

ref: #732 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/server/basket/helpers.go
Comment (aaronc) [inline x/ecocredit/server/basket/helpers.go:1]: if we're copying this from `data_prefixes.go` we should just move it to `keys.go` and share it between the two packages

File: x/ecocredit/server/basket/put.go
Comment (technicallyty) [reply x/ecocredit/server/basket/put.go]: yeah, didnt know what else to put really.. changed it to just `calculateTokens`

File: types/math/dec.go
Comment (aaronc) [inline types/math/dec.go:93]: docs

File: types/math/dec.go
Comment (aaronc) [inline types/math/dec.go:123]: docs

File: x/ecocredit/server/basket/put.go
Comment (aaronc) [inline x/ecocredit/server/basket/put.go:170]: From my PR pretty sure this should be:
```suggestion
	multiplier := regenmath.NewDecFinite(1, int32(exp))
```

**Review by aaronc — COMMENTED** [review]
Good work @technicallyty 👍

Took a quick look, will look more later when I'm better rested

Let's change usages of `award` to `mint`

File: types/math/dec.go
Comment (technicallyty) [reply types/math/dec.go:125]: added

File: types/math/dec.go
Comment (technicallyty) [reply types/math/dec.go:94]: added

File: x/ecocredit/server/basket/put.go
Comment (technicallyty) [reply x/ecocredit/server/basket/put.go]: yup, fixed

File: types/math/dec.go
Comment (aaronc) [inline types/math/dec.go:124]:
```suggestion
// MulExact returns a new dec with value x * y. The product must not round or an error will be returned.
```

File: x/ecocredit/server/basket/put.go
Comment (aaronc) [inline x/ecocredit/server/basket/put.go:20]: must check error

File: x/ecocredit/server/basket/put.go
Comment (aaronc) [inline x/ecocredit/server/basket/put.go:74]:
```suggestion
func (k Keeper) canBasketAcceptCredit(ctx context.Context, basket *basketv1.Basket, batchInfo *ecocredit.BatchInfo) error {
```

File: x/ecocredit/server/basket/put.go
Comment (aaronc) [inline x/ecocredit/server/basket/put.go:120]:
```suggestion
func (k Keeper) transferToBasket(ctx context.Context, sender sdk.AccAddress, amt regenmath.Dec, basket *basketv1.Basket, batchInfo *ecocredit.BatchInfo) error {
```

File: x/ecocredit/server/basket/put.go
Comment (aaronc) [inline x/ecocredit/server/basket/put.go:65]: we can do this outside the loop

File: x/ecocredit/server/basket/put.go
Comment (aaronc) [inline x/ecocredit/server/basket/put.go:81]:
```suggestion
		switch criteria := basket.DateCriteria.Sum.(type) {
```

File: x/ecocredit/server/basket/put.go
Comment (aaronc) [inline x/ecocredit/server/basket/put.go:87]:
```suggestion
			minStartDate = blockTime.Sub(window)
```

File: x/ecocredit/server/basket/put.go
Comment (aaronc) [inline x/ecocredit/server/basket/put.go:93]:
```suggestion
			return sdkerrors.ErrInvalidRequest.Wrapf("cannot put a credit from a batch with start date %s "+
				"into a basket that requires an earliest start date of %s", batchInfo.StartDate.String(), minStartDate.String())
		}
```

File: x/ecocredit/server/basket/put.go
Comment (aaronc) [inline x/ecocredit/server/basket/put.go:168]:
```suggestion
func creditAmountToBasketCoins(creditAmt regenmath.Dec, exp uint32, denom string) (sdk.Coins, error) {
```

File: x/ecocredit/server/basket/put_test.go
Comment (aaronc) [inline x/ecocredit/server/basket/put_test.go:92]:
```suggestion
	var dur time.Duration = 500000000000000000
```

pretty sure this should be positive (I had another comment to switch `Add` to `Sub` )

File: x/ecocredit/server/basket/put_test.go
Comment (aaronc) [inline x/ecocredit/server/basket/put_test.go:328]: we should check the expected credit balance after adding to the basket

File: x/ecocredit/server/invariants.go
Comment (aaronc) [inline x/ecocredit/server/invariants.go:29]: We actually need to update invariants, because now `tradable supply = SUM(user balances) + SUM(basket balances)`

**Review by aaronc — COMMENTED** [review]
This is looking very close! Just a few things to update and then we can merge 🎉

File: types/math/dec.go
Comment (robert-zaremba) [reply types/math/dec.go:94]: `(10 * coeff)^exp` looks weird.   `NewDecFinite(3, 4) = 30^4 =810000`

File: x/ecocredit/server/basket/put.go
Comment (technicallyty) [reply x/ecocredit/server/basket/put.go:20]: wouldn't that get caught in validate basic?

File: x/ecocredit/server/basket/put.go
Comment (technicallyty) [reply x/ecocredit/server/basket/put.go:87]: Sub is an operation for time.Time, it doesn't accept duration

File: x/ecocredit/server/invariants.go
Comment (technicallyty) [reply x/ecocredit/server/invariants.go:29]: I'm not sure im following - if its just sending credits from one entity to another, doesn't the supply just stay the same? not new credits are being created?

Comment (aaronc) [reply x/ecocredit/server/invariants.go:29]: yes, but the invariant checks that all the balances = the supply. and now the invariant check will fail because we need to check basket balances too

Comment (technicallyty) [reply x/ecocredit/server/invariants.go:29]: oh okay, i thought for a second you meant to change the TradableSupply in the store 😨 i'll update this function

File: x/ecocredit/server/basket/put.go
Comment (aaronc) [reply x/ecocredit/server/basket/put.go:20]: i guess but still...

File: x/ecocredit/server/invariants.go
Comment (technicallyty) [reply x/ecocredit/server/invariants.go:61]: added basket check to invariants. feels kinda hacky though, let me know how it looks

File: types/math/dec.go
Comment (aaronc) [inline types/math/dec.go:93]:
```suggestion
// NewDecFinite returns a decimal with a value of coeff * 10^exp.
```

**Review by aaronc — APPROVED** [review]
LGTM. Great work @technicallyty 👏

**Review by ryanchristo — COMMENTED** [review]
Very nice! A few comments and `NewKeeper` needs to be updated in a couple tests following the latest merge.

File: x/ecocredit/server/basket/put.go
Comment (technicallyty) [reply x/ecocredit/server/basket/put.go:21]: added

File: x/ecocredit/keys.go
Comment (technicallyty) [reply x/ecocredit/keys.go:75]: added to all un-godoc'd functions

File: x/ecocredit/server/basket/put_test.go
Comment (technicallyty) [reply x/ecocredit/server/basket/put_test.go]: changed em 👍🏻

**Comment by technicallyty** [comment]
conflict with https://github.com/regen-network/regen-ledger/pull/765 - maybe we should wait on this PR til that gets merged

**Review by robert-zaremba — COMMENTED** [review]
pre-approving, but before merging let's consider:
* using subaccount for baskets
* moving decimal functions to a separate package
*

File: x/ecocredit/server/basket/put.go
Comment (aaronc) [reply x/ecocredit/server/basket/put.go:71]: we just need to make sure we put this in app.go

File: x/ecocredit/keys.go
Comment (aaronc) [reply x/ecocredit/keys.go:100]: this is mostly code that was already there and will be going way soon in #723 so I don't think we really need to do much cleanup

File: x/ecocredit/server/basket/put.go
Comment (technicallyty) [reply x/ecocredit/server/basket/put.go:71]: its not really an address though, how about BasketModuleName?

Comment (aaronc) [reply x/ecocredit/server/basket/put.go:71]: `BasketSubModuleName`

File: x/ecocredit/keys.go
Comment (technicallyty) [reply x/ecocredit/keys.go:100]: perhaps we can just do that when/if the times comes when another package needs it? this move was just a fix for import cycles

Comment (aaronc) [reply x/ecocredit/keys.go:100]: yeah i'd say keep as is because it's all getting deleted right after the 2.2 release

File: x/ecocredit/keys.go
Comment (technicallyty) [reply x/ecocredit/keys.go:46]: added comment explaining format

**Review by ryanchristo — APPROVED** [review]

**Comment by ryanchristo** [comment]
Aside from failing test and go mod tidy, looks good to me.

**Comment by aaronc** [comment]
> Aside from failing test and go mod tidy, looks good to me.

what's the failing test?

**Comment by aaronc** [comment]
> > Aside from failing test and go mod tidy, looks good to me.
> 
> what's the failing test?

ah i see, but that's unrelated. let's push a fix separately?

**Pull request merged**
